PR: https://github.com/odoo/odoo/pull/217895

From: 77f860f5d3757e5a56861ac1de95b9ad29ea0dff
From: Abdelrahman Mahmoud (amah)
Date: 2025-08-29 09:56:43

Breaking data model changes scores: del:3.4 + add:0, change matches:
-    date_from = fields.Date(string='Starting Date')
-    date_to = fields.Date(string='End Date')
-    duration_hours = fields.Float(compute='_compute_duration_hours', string='Duration (hours)')
+    duration_hours = fields.Float(compute='_compute_duration_hours', inverse='_inverse_duration_hours', string='Duration (hours)', store=True, readonly=False)
-    resource_id = fields.Many2one('resource.resource', 'Resource')

Total Changes: 179

[IMP] resource: Resource attendance rework

* Added 'full_day' as a selection value for `day_period`
* Displayed `duration_hours` in the list view
* Made `duration_hours` editable and compute `hours` based on it
* Removed `date_from` and `date_to` fields for simplification

closes odoo/odoo#217895

Task: 4929940
Related: odoo/enterprise#92500
Related: odoo/upgrade#8030
Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>

================================= pseudo patch: =================================

--- a/addons/resource/models/resource_calendar.py
+++ b/addons/resource/models/resource_calendar.py
@@ -1,6 +1,5 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-import itertools
 from collections import defaultdict
 from datetime import datetime, time, timedelta
 from functools import partial
@@ -87,6 +86,7 @@ class ResourceCalendar(models.Model):
             "- Flexible : Define an amount of hours to work on the week.\n"
             "- Fully Fixed : define the days, periods and the start & end time for each period of the day",
     )
+    duration_based = fields.Boolean("Attendance based on duration", help="The hours will be centered around 12:00 to cover the duration for the day")
     flexible_hours = fields.Boolean(string="Flexible Hours",
         compute="_compute_flexible_hours", inverse="_inverse_flexible_hours", store=True,
         help="When enabled, it will allow employees to work flexibly, without relying on the company's working schedule (working hours).")
@@ -130,7 +130,7 @@ class ResourceCalendar(models.Model):
 
             # Avoid superimpose in attendance
             attendance_ids = res_calendar.attendance_ids.filtered(
-                lambda attendance: not attendance.resource_id and attendance.display_type is False)
+                lambda attendance: not attendance.display_type)
             if res_calendar.two_weeks_calendar:
                 res_calendar._check_overlap(attendance_ids.filtered(lambda attendance: attendance.week_type == '0'))
                 res_calendar._check_overlap(attendance_ids.filtered(lambda attendance: attendance.week_type == '1'))
@@ -177,7 +177,7 @@ class ResourceCalendar(models.Model):
                 'two_weeks_calendar': company_calendar.two_weeks_calendar,
                 'tz': company_calendar.tz,
                 'attendance_ids': [(5, 0, 0)] + [
-                    (0, 0, attendance._copy_attendance_vals()) for attendance in company_calendar.attendance_ids if not attendance.resource_id],
+                    (0, 0, attendance._copy_attendance_vals()) for attendance in company_calendar.attendance_ids],
             })
 
     @api.onchange('attendance_ids')
@@ -287,37 +287,26 @@ class ResourceCalendar(models.Model):
         self.ensure_one()
         if not self.two_weeks_calendar:
             self.two_weeks_calendar = True
-            final_attendances = [
-                Command.create({
-                    'name': 'First week',
-                    'dayofweek': '0',
-                    'sequence': '0',
-                    'hour_from': 0,
-                    'day_period': 'morning',
-                    'week_type': '0',
-                    'hour_to': 0,
-                    'display_type':
-                    'line_section'}),
-                Command.create({
-                    'name': 'Second week',
-                    'dayofweek': '0',
-                    'sequence': '25',
-                    'hour_from': 0,
-                    'day_period': 'morning',
-                    'week_type': '1',
-                    'hour_to': 0,
-                    'display_type': 'line_section'}),
-            ]
-            for idx, att in enumerate(self.attendance_ids):
-                final_attendances.append(Command.create(dict(att._copy_attendance_vals(), week_type='0', sequence=idx + 1)))
-                final_attendances.append(Command.create(dict(att._copy_attendance_vals(), week_type='1', sequence=idx + 26)))
+            final_attendances = self._get_two_weeks_attendance()
             self.attendance_ids = [Command.clear()] + final_attendances
 
         else:
             self.two_weeks_calendar = False
             self.attendance_ids.unlink()
+            self.duration_based = False
             self.attendance_ids = self._get_default_attendance_ids(self.company_id)
 
+    def switch_based_on_duration(self):
+        self.ensure_one()
+        self.duration_based = not self.duration_based
+        if self.duration_based:
+            self.attendance_ids.filtered(lambda att: att.day_period == 'lunch').unlink()
+        else:
+            self.attendance_ids.unlink()
+            self.attendance_ids = self._get_default_attendance_ids(self.company_id)
+            if self.two_weeks_calendar:
+                self.attendance_ids = self._get_two_weeks_attendance()
+
     # --------------------------------------------------
     # Computation API
     # --------------------------------------------------
@@ -330,11 +319,9 @@ class ResourceCalendar(models.Model):
             resources_list = [resources]
         else:
             resources_list = list(resources) + [self.env['resource.resource']]
-        resource_ids = [r.id for r in resources_list]
         domain = Domain.AND([
             Domain(domain or Domain.TRUE),
             Domain('calendar_id', '=', self.id),
-            Domain('resource_id', 'in', resource_ids),
             Domain('display_type', '=', False),
             Domain('day_period', '!=' if not lunch else '=', 'lunch'),
         ])
@@ -346,14 +333,11 @@ class ResourceCalendar(models.Model):
         for resource in resources_list:
             resources_per_tz[tz or timezone((resource or self).tz)].append(resource)
         # Resource specific attendances
-        attendance_per_resource = defaultdict(lambda: self.env['resource.calendar.attendance'])
         # Calendar attendances per day of the week
         # * 7 days per week * 2 for two week calendars
         attendances_per_day = [self.env['resource.calendar.attendance']] * 7 * 2
         weekdays = set()
         for attendance in attendances:
-            if attendance.resource_id:
-                attendance_per_resource[attendance.resource_id] |= attendance
             weekday = int(attendance.dayofweek)
             weekdays.add(weekday)
             if self.two_weeks_calendar:
@@ -377,20 +361,13 @@ class ResourceCalendar(models.Model):
         days = rrule(DAILY, start.date(), until=end.date(), byweekday=weekdays)
         ResourceCalendarAttendance = self.env['resource.calendar.attendance']
         base_result = []
-        per_resource_result = defaultdict(list)
         for day in days:
             week_type = ResourceCalendarAttendance.get_week_type(day)
             attendances = attendances_per_day[day.weekday() + 7 * week_type]
             for attendance in attendances:
-                if (attendance.date_from and day.date() < attendance.date_from) or\
-                    (attendance.date_to and attendance.date_to < day.date()):
-                    continue
                 day_from = datetime.combine(day, float_to_time(attendance.hour_from))
                 day_to = datetime.combine(day, float_to_time(attendance.hour_to))
-                if attendance.resource_id:
-                    per_resource_result[attendance.resource_id].append((day_from, day_to, attendance))
-                else:
-                    base_result.append((day_from, day_to, attendance))
+                base_result.append((day_from, day_to, attendance))
 
         # Copy the result localized once per necessary timezone
         # Strictly speaking comparing start_dt < time or start_dt.astimezone(tz) < time
@@ -473,9 +450,6 @@ class ResourceCalendar(models.Model):
                         current_start_day += timedelta(days=7)
 
                     result_per_resource_id[resource.id] = Intervals(intervals, keep_distinct=True)
-                elif resource in per_resource_result:
-                    resource_specific_result = [(max(bounds_per_tz[tz][0], tz.localize(val[0])), min(bounds_per_tz[tz][1], tz.localize(val[1])), val[2]) for val in per_resource_result[resource]]
-                    result_per_resource_id[resource.id] = Intervals(itertools.chain(res, resource_specific_result), keep_distinct=True)
                 else:
                     result_per_resource_id[resource.id] = res_intervals
         return result_per_resource_id
@@ -605,7 +579,7 @@ class ResourceCalendar(models.Model):
         """ attendance_ids correspond to attendance of a week,
             will check for each day of week that there are no superimpose. """
         result = []
-        for attendance in attendance_ids.filtered(lambda att: not att.date_from and not att.date_to):
+        for attendance in attendance_ids:
             # 0.000001 is added to each start hour to avoid to detect two contiguous intervals as superimposing.
             # Indeed Intervals function will join 2 intervals with the start and stop hour corresponding.
             result.append((int(attendance.dayofweek) * 24 + attendance.hour_from + 0.000001, int(attendance.dayofweek) * 24 + attendance.hour_to, attendance))
@@ -704,8 +678,7 @@ class ResourceCalendar(models.Model):
     def _get_global_attendances(self):
         return self.attendance_ids.filtered(lambda attendance:
             attendance.day_period != 'lunch'
-            and not attendance.date_from and not attendance.date_to
-            and not attendance.resource_id and not attendance.display_type)
+            and not attendance.display_type)
 
     def _get_unusual_days(self, start_dt, end_dt, company_id=False):
         if not self:
@@ -736,8 +709,6 @@ class ResourceCalendar(models.Model):
                     'hour_from': attendance.hour_from,
                     'hour_to': attendance.hour_to,
                     'day_period': attendance.day_period,
-                    'date_from': attendance.date_from,
-                    'date_to': attendance.date_to,
                     'display_type': attendance.display_type,
                 })
                 for attendance in attendances
@@ -760,6 +731,32 @@ class ResourceCalendar(models.Model):
             Command.create({'name': self.env._('Friday Afternoon'), 'dayofweek': '4', 'hour_from': 13, 'hour_to': 17, 'day_period': 'afternoon'}),
         ]
 
+    def _get_two_weeks_attendance(self):
+        final_attendances = [
+            Command.create({
+                'name': 'First week',
+                'dayofweek': '0',
+                'sequence': '0',
+                'hour_from': 0,
+                'day_period': 'morning',
+                'week_type': '0',
+                'hour_to': 0,
+                'display_type':
+                'line_section'}),
+            Command.create({
+                'name': 'Second week',
+                'dayofweek': '0',
+                'sequence': '25',
+                'hour_from': 0,
+                'day_period': 'morning',
+                'week_type': '1',
+                'hour_to': 0,
+                'display_type': 'line_section'}),
+        ]
+        for idx, att in enumerate(self.attendance_ids):
+            final_attendances.append(Command.create(dict(att._copy_attendance_vals(), week_type='0', sequence=idx + 1)))
+            final_attendances.append(Command.create(dict(att._copy_attendance_vals(), week_type='1', sequence=idx + 26)))
+        return final_attendances
     # --------------------------------------------------
     # External API
     # --------------------------------------------------
@@ -937,16 +934,25 @@ class ResourceCalendar(models.Model):
             ('display_type', '=', False),
             ('day_period', '!=', 'lunch'),
         ]
-        if day_period:
-            domain.append(('day_period', '=', day_period))
 
-        attendances = self.env['resource.calendar.attendance']._read_group(domain=domain,
-            groupby=['week_type', 'dayofweek'],
-            aggregates=['hour_from:min', 'hour_to:max'],
-            order='dayofweek')
+        init_attendances = self.env['resource.calendar.attendance']._read_group(domain=domain,
+        groupby=['week_type', 'dayofweek', 'day_period'],
+        aggregates=['hour_from:min', 'hour_to:max'],
+        order='dayofweek,hour_from:min')
 
-        attendances = [DummyAttendance(hour_from, hour_to, dayofweek, None, week_type)
-            for week_type, dayofweek, hour_from, hour_to in attendances]
+        init_attendances = [DummyAttendance(hour_from, hour_to, dayofweek, day_period, week_type)
+            for week_type, dayofweek, day_period, hour_from, hour_to in init_attendances]
+
+        if day_period:
+            attendances = [att for att in init_attendances if att.day_period == day_period]
+            for attendance in filter(lambda att: att.day_period == 'full_day', init_attendances):
+                attendances.append(attendance._replace(
+                    hour_from=attendance.hour_from if day_period == 'morning' else 12,
+                    hour_to=attendance.hour_to if day_period == 'afternoon' else 12,
+                ))
+
+        else:
+            attendances = init_attendances
 
         default_start = min((att.hour_from for att in attendances), default=0.0)
         default_end = max((att.hour_to for att in attendances), default=0.0)
@@ -955,12 +961,9 @@ class ResourceCalendar(models.Model):
         if self.two_weeks_calendar:
             week_type = str(self.env['resource.calendar.attendance'].get_week_type(target_date))
 
-        hour_from = next(
-            (att.hour_from for att in attendances if int(att.dayofweek) == target_date.weekday() and
-            (att.week_type == week_type)), default_start)
-        hour_to = next(
-            (att.hour_to for att in attendances if int(att.dayofweek) == target_date.weekday() and
-            (att.week_type == week_type)), default_end)
+        filtered_attendances = [att for att in attendances if att.week_type == week_type and int(att.dayofweek) == target_date.weekday()]
+        hour_from = min((att.hour_from for att in filtered_attendances), default=default_start)
+        hour_to = max((att.hour_to for att in filtered_attendances), default=default_end)
 
         return (hour_from, hour_to)
 

--- a/addons/resource/models/resource_calendar_attendance.py
+++ b/addons/resource/models/resource_calendar_attendance.py
@@ -2,7 +2,8 @@
 
 import math
 
-from odoo import api, fields, models, _
+from odoo import api, fields, models
+from odoo.exceptions import UserError
 
 
 class ResourceCalendarAttendance(models.Model):
@@ -20,23 +21,22 @@ class ResourceCalendarAttendance(models.Model):
         ('5', 'Saturday'),
         ('6', 'Sunday')
         ], 'Day of Week', required=True, index=True, default='0')
-    date_from = fields.Date(string='Starting Date')
-    date_to = fields.Date(string='End Date')
-    hour_from = fields.Float(string='Work from', required=True, index=True,
+    hour_from = fields.Float(string='Work from', default=0, required=True, index=True,
         help="Start and End time of working.\n"
              "A specific value of 24:00 is interpreted as 23:59:59.999999.")
-    hour_to = fields.Float(string='Work to', required=True)
+    hour_to = fields.Float(string='Work to', default=0, required=True)
     # For the hour duration, the compute function is used to compute the value
     # unambiguously, while the duration in days is computed for the default
     # value based on the day_period but can be manually overridden.
-    duration_hours = fields.Float(compute='_compute_duration_hours', string='Duration (hours)')
+    duration_hours = fields.Float(compute='_compute_duration_hours', inverse='_inverse_duration_hours', string='Duration (hours)', store=True, readonly=False)
     duration_days = fields.Float(compute='_compute_duration_days', string='Duration (days)', store=True, readonly=False)
     calendar_id = fields.Many2one("resource.calendar", string="Resource's Calendar", required=True, index=True, ondelete='cascade')
+    duration_based = fields.Boolean(related="calendar_id.duration_based")
     day_period = fields.Selection([
         ('morning', 'Morning'),
         ('lunch', 'Break'),
-        ('afternoon', 'Afternoon')], required=True, default='morning')
-    resource_id = fields.Many2one('resource.resource', 'Resource')
+        ('afternoon', 'Afternoon'),
+        ('full_day', 'Full Day')], required=True, default='morning')
     week_type = fields.Selection([
         ('1', 'Second'),
         ('0', 'First')
@@ -58,6 +58,12 @@ class ResourceCalendarAttendance(models.Model):
         # avoid wrong order
         self.hour_to = max(self.hour_to, self.hour_from)
 
+    @api.constrains('day_period')
+    def _check_day_period(self):
+        for attendance in self:
+            if attendance.day_period == 'lunch' and attendance.duration_based:
+                raise UserError(self.env._("%(att)s is a break attendance, You should not have such record on duration based calendar", att=attendance.name))
+
     @api.model
     def get_week_type(self, date):
         # week_type is defined by
@@ -70,14 +76,32 @@ class ResourceCalendarAttendance(models.Model):
 
     @api.depends('hour_from', 'hour_to')
     def _compute_duration_hours(self):
-        for attendance in self:
+        for attendance in self.filtered('hour_to'):
             attendance.duration_hours = (attendance.hour_to - attendance.hour_from) if attendance.day_period != 'lunch' else 0
 
-    @api.depends('day_period', 'duration_hours')
+    def _inverse_duration_hours(self):
+        for calendar, attendances in self.grouped('calendar_id').items():
+            if not calendar.duration_based:
+                continue
+            for attendance in attendances:
+                if attendance.day_period == 'full_day':
+                    period_duration = attendance.duration_hours / 2
+                    attendance.hour_to = 12 + period_duration
+                    attendance.hour_from = 12 - period_duration
+                elif attendance.day_period == 'morning':
+                    attendance.hour_to = 12
+                    attendance.hour_from = 12 - attendance.duration_hours
+                elif attendance.day_period == 'afternoon':
+                    attendance.hour_to = 12 + attendance.duration_hours
+                    attendance.hour_from = 12
+
+    @api.depends('day_period')
     def _compute_duration_days(self):
         for attendance in self:
             if attendance.day_period == 'lunch':
                 attendance.duration_days = 0
+            elif attendance.day_period == 'full_day':
+                attendance.duration_days = 1
             else:
                 attendance.duration_days = 0.5 if attendance.duration_hours <= attendance.calendar_id.hours_per_day * 3 / 4 else 1
 
@@ -85,8 +109,8 @@ class ResourceCalendarAttendance(models.Model):
     def _compute_display_name(self):
         super()._compute_display_name()
         this_week_type = str(self.get_week_type(fields.Date.context_today(self)))
-        section_names = {'0': _('First week'), '1': _('Second week')}
-        section_info = {True: _('this week'), False: _('other week')}
+        section_names = {'0': self.env._('First week'), '1': self.env._('Second week')}
+        section_info = {True: self.env._('this week'), False: self.env._('other week')}
         for record in self.filtered(lambda l: l.display_type == 'line_section'):
             section_name = f"{section_names[record.week_type]} ({section_info[this_week_type == record.week_type]})"
             record.display_name = section_name
@@ -96,8 +120,6 @@ class ResourceCalendarAttendance(models.Model):
         return {
             'name': self.name,
             'dayofweek': self.dayofweek,
-            'date_from': self.date_from,
-            'date_to': self.date_to,
             'hour_from': self.hour_from,
             'hour_to': self.hour_to,
             'day_period': self.day_period,
@@ -107,4 +129,4 @@ class ResourceCalendarAttendance(models.Model):
         }
 
     def _is_work_period(self):
-        return self.day_period != 'lunch'
+        return self.day_period != 'lunch' and not self.display_type
